ch_pipeline/hfb/io.py

```diff
@@ -0,0 +1,178 @@
+"""HFB tasks for reading and writing files
+"""
+
+import gc
+import numpy as np
+
+from caput import pipeline
+from caput import config
+
+from ch_util import ephemeris
+
+from draco.core import io
+
+from beam_model.formed import FFTFormedActualBeamModel
+
+from .containers import HFBReader
+
+
+class LoadHFBDataFiles(io.BaseLoadFiles):
+    """Load CHIME HFB data from a file list passed into the setup routine.
+
+    Attributes
+    ----------
+    source_dec : float
+        Declination of source in degrees.
+    beam_ew_exclude : list
+        List of East-West beam indices (i.e., in the range 0-3) to exclude.
+    freq_phys_range : list
+        Start and stop of physical frequencies (in MHz) to read. The mean is
+        used as reference frequency in evaluating beam positions (for selecting
+        the beams closest to a transiting source).
+    freq_phys_list : list
+        List of physical frequencies (in MHz) to read. The first frequency
+        in this list is also used in evaluating beam positions (for selecting
+        the beams closest to a transiting source).
+
+    Selections
+    ----------
+    Selections in frequency and beams can be done in two ways:
+    1. By passing a `source_dec` attribute (for the beam selection) and/or
+       a `freq_phys_range` or `freq_phys_list` attribute (for the frequency
+       selection). If both `freq_phys_range` and `freq_phys_list` are given the
+       former will take precedence, but you should clearly avoid doing this.
+    2. By manually passing indices in the `selections` attribute
+       (see documentation in :class:`draco.core.io.BaseLoadFiles`).
+    Method 1 takes precedence over method 2. If no relevant attributes are
+    passed, all frequencies/beams are read.
+    """
+
+    files = None
+
+    _file_ptr = 0
+
+    source_dec = config.Property(proptype=float, default=None)
+    beam_ew_exclude = config.Property(proptype=list, default=[])
+    freq_phys_list = config.Property(proptype=list, default=[])
+    freq_phys_range = config.Property(proptype=list, default=[])
+
+    def setup(self, files):
```

Comment:
I think we'll also want to support loading files specified in the config file. You may just need to derive a new class from this like all the variants in the main io tasks.

Reply:
I now created `BaseLoadHFBDataFiles` and `LoadHFBDataFromFilelist`.

Reply:
It seems we cannot use the `_list_of_filelists`, `_list_or_glob`, and `_list_of_filegroups` from `draco.core.io` as they are: currently we pass along a `time_range` for each group of files to only load data around transit time.

The HFB data loading was build around the `HFBReader(caput.tod.Reader)`, and was based on `ch_pipeline.core.io.LoadDataFiles`, which seems to be a different way of loading data, coexisting with the stuff found in `draco.core.io`.

Not sure how to proceed here:
1. Remove `HFBReader` and remake the data loading, more like `draco.core.io`.
2. Keep `HFBReader` and keep things more the way they are, but add some way of loading files specified in the config file.

Reply:
Went with option 2 in the above comment.

The structure in `ch_pipeline.hfb.io` now follows that of `draco.core.io` as much as possible:
1. The loading of files is handled by the method `_load_filelist` in `hfb.io.BaseLoadFiles`, which calls `hfb.containers.HFBReader`.
2. To load files specified in the config file (potentially with time ranges): use `hfb.io.LoadFilesFromParams`, which expects `filegroups` and processes them by calling `_load_filelist` (example given in documentation).
3. To load files specified by `core.dataquery.QueryDatabase`: use `hfb.io.LoadFiles`, which converts the output of the `QueryDatabase` to `filegroups`, so they can be processed by `LoadFilesFromParams`.